service/controller/resource/alerting/alertmanagerconfig/create.go

```diff
@@ -0,0 +1,39 @@
+package alertmanagerconfig
+
+import (
+	"context"
+
+	"github.com/giantswarm/microerror"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/giantswarm/prometheus-meta-operator/v2/service/controller/resource/resourceutils"
+)
+
+func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+	desired, err := r.toSecret()
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	r.config.Logger.Debugf(ctx, "creating")
+	current, err := r.config.K8sClient.K8sClient().CoreV1().Secrets(desired.GetNamespace()).Get(ctx, desired.GetName(), metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		current, err = r.config.K8sClient.K8sClient().CoreV1().Secrets(desired.GetNamespace()).Create(ctx, desired, metav1.CreateOptions{})
+	}
+
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	if r.hasChanged(current, desired) {
+		resourceutils.UpdateMeta(current, desired)
+		_, err = r.config.K8sClient.K8sClient().CoreV1().Secrets(desired.GetNamespace()).Update(ctx, desired, metav1.UpdateOptions{})
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+	r.config.Logger.Debugf(ctx, "created")
```

Comment:
I really dislike this kind of logic being duplicated everywhere.

Reply:
Well, I never really liked the generic resource so :D

Also, this is going away. The main idea is that we can start disabling things slowly when migrating to mimir (especially the clean up part) so that we can add  the pseudo algorigthm in the ensure create resources that we want ;)

```go
if mimirEnabled {
  return EnsureDelete()
}
```

Reply:
I get that, but it's a lot of code duplication and its very error prone.

Reply:
I understand your concern